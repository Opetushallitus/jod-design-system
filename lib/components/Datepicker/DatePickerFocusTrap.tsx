import { FocusTrap } from 'focus-trap-react';

export const DatepickerFocusTrap = ({ children, active }: { children: React.ReactNode; active: boolean }) => {
  return (
    <FocusTrap
      active={active}
      focusTrapOptions={{
        initialFocus: false,
        escapeDeactivates: true,
        clickOutsideDeactivates: true,
        checkCanFocusTrap: async (containers) => {
          return new Promise((resolve) => {
            let tries = 0;
            const check = () => {
              const el = containers[0].querySelector('[data-scope="date-picker"][data-part="view-control"]') as
                | HTMLDivElement
                | undefined;

              if (el || tries > 10) {
                resolve();
              } else {
                setTimeout(check, 10);
                tries += 1;
              }
            };
            check();
          });
        },
      }}
    >
      {children}
    </FocusTrap>
  );
};
