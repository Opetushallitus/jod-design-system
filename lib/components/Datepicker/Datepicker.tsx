import { DatePicker as ArkDatePicker, Portal, UseDatePickerContext } from '@ark-ui/react';
import React from 'react';
import { cx } from '../../cva';

const MIN_YEAR = 1900;
const MAX_YEAR = 2100;
const isInvalidYear = (year: number) => year < MIN_YEAR || year >= MAX_YEAR;

const tableCellClasses =
  'ds-m-3 ds-text-body-sm ds-size-[28px] sm:ds-size-[32px] hover:ds-underline focus:ds-left-auto ds-capitalize data-[selected]:ds-bg-secondary-1-50 data-[selected]:ds-rounded-full';

const getDayCellClasses = (datePicker: UseDatePickerContext, day: UseDatePickerContext['focusedValue']) =>
  cx(tableCellClasses, {
    'disabled:ds-text-inactive-gray disabled:ds-cursor-not-allowed': datePicker.isUnavailable(day),
    'ds-text-inactive-gray': datePicker.focusedValue.month !== day.month,
  });

const getYearCellClasses = (year: { label: string; value: number }) =>
  cx(tableCellClasses, {
    'disabled:ds-text-inactive-gray disabled:ds-cursor-not-allowed': isInvalidYear(year.value),
  });

const Header = () => (
  <ArkDatePicker.ViewControl className="ds-mb-4 ds-flex ds-justify-between">
    <ArkDatePicker.PrevTrigger className="material-symbols-outlined ds-text-accent ds-font-semibold ds-px-3">
      arrow_back
    </ArkDatePicker.PrevTrigger>

    <ArkDatePicker.ViewTrigger>
      <ArkDatePicker.RangeText className="ds-capitalize ds-font-bold" />
    </ArkDatePicker.ViewTrigger>

    <ArkDatePicker.NextTrigger className="material-symbols-outlined ds-text-accent ds-font-semibold ds-px-3">
      arrow_forward
    </ArkDatePicker.NextTrigger>
  </ArkDatePicker.ViewControl>
);

export interface DatepickerProps {
  /** Name of the input field */
  name?: string;
  /** Initial value, the internal format is 'yyyy-mm-dd' */
  value?: string;
  /** Callback for when the input field loses focus */
  onBlur?: (event: React.FocusEvent<HTMLInputElement>) => void;
  /** Callback for when the value of the input field changes */
  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
  /** Label text for input */
  label: string;
  /** Placeholder text, eg. "pp.kk.vvvv" */
  placeholder?: string;
  /** Help text to display below the input field */
  help?: string;
}

/** Datepicker component for selecting a date. */
export const Datepicker = React.forwardRef<HTMLInputElement, DatepickerProps>(function Datepicker(
  { value, name, label, placeholder, help, onBlur, onChange }: DatepickerProps,
  ref,
) {
  const helpId = React.useId();
  return (
    <ArkDatePicker.Root
      onValueChange={(details) => {
        onChange({
          target: { name, value: details.valueAsString[0] ?? '' },
        } as React.ChangeEvent<HTMLInputElement>);
      }}
      value={value ? [value] : []}
      locale="fi-FI"
      timeZone="Europe/Helsinki"
      className="ds-w-full"
      isDateUnavailable={(date) => isInvalidYear(date.year)}
    >
      <ArkDatePicker.Label className="ds-mb-4 ds-inline-block ds-align-top ds-text-form-label ds-text-black">
        {label}
      </ArkDatePicker.Label>
      <ArkDatePicker.Control>
        <div className="ds-flex ds-w-full">
          <ArkDatePicker.Input
            ref={ref}
            name={name}
            placeholder={placeholder}
            className="ds-w-full ds-rounded-l ds-border-y ds-border-l ds-border-border-gray ds-bg-white ds-p-5 ds-text-black ds-outline-none placeholder:ds-text-inactive-gray placeholder:ds-font-poppins"
            onInput={(e: React.ChangeEvent<HTMLInputElement>) => {
              if (e.target.value === '') {
                onChange({
                  target: { name, value: '' },
                } as React.ChangeEvent<HTMLInputElement>);
              }
            }}
            onBlur={onBlur}
          />
          <ArkDatePicker.Trigger className="material-symbols-outlined ds-select-none ds-rounded-r ds-border-y ds-border-r ds-border-border-gray ds-bg-white ds-p-5 ds-text-secondary-gray">
            calendar_month
          </ArkDatePicker.Trigger>
        </div>
      </ArkDatePicker.Control>
      {help && (
        <div id={helpId} className="ds-mt-2 ds-block ds-text-help ds-text-secondary-gray">
          {help}
        </div>
      )}
      <Portal>
        <ArkDatePicker.Positioner>
          <ArkDatePicker.Content className="ds-z-50 ds-rounded ds-shadow-border ds-bg-white ds-p-4">
            <ArkDatePicker.View view="day">
              <ArkDatePicker.Context>
                {(datePicker) => (
                  <>
                    <Header />
                    <ArkDatePicker.Table>
                      <ArkDatePicker.TableHead>
                        <ArkDatePicker.TableRow>
                          {datePicker.weekDays.map((weekDay) => (
                            <ArkDatePicker.TableHeader
                              key={weekDay.long}
                              className="ds-capitalize ds-size-[28px] sm:ds-size-[32px]"
                            >
                              {weekDay.short}
                            </ArkDatePicker.TableHeader>
                          ))}
                        </ArkDatePicker.TableRow>
                      </ArkDatePicker.TableHead>
                      <ArkDatePicker.TableBody>
                        {datePicker.weeks.map((week, i) => (
                          <ArkDatePicker.TableRow key={i}>
                            {week.map((day) => (
                              <ArkDatePicker.TableCell key={`${day.day}_${i}`} value={day}>
                                <ArkDatePicker.TableCellTrigger
                                  className={getDayCellClasses(datePicker, day)}
                                  disabled={datePicker.isUnavailable(day)}
                                >
                                  {day.day}
                                </ArkDatePicker.TableCellTrigger>
                              </ArkDatePicker.TableCell>
                            ))}
                          </ArkDatePicker.TableRow>
                        ))}
                      </ArkDatePicker.TableBody>
                    </ArkDatePicker.Table>
                  </>
                )}
              </ArkDatePicker.Context>
            </ArkDatePicker.View>
            <ArkDatePicker.View view="month">
              <ArkDatePicker.Context>
                {(datePicker) => (
                  <>
                    <Header />
                    <ArkDatePicker.Table>
                      <ArkDatePicker.TableBody>
                        {datePicker.getMonthsGrid({ columns: 4, format: 'short' }).map((months, i) => (
                          <ArkDatePicker.TableRow key={i}>
                            {months.map((month) => (
                              <ArkDatePicker.TableCell key={month.label} value={month.value}>
                                <ArkDatePicker.TableCellTrigger className={tableCellClasses}>
                                  {month.label}
                                </ArkDatePicker.TableCellTrigger>
                              </ArkDatePicker.TableCell>
                            ))}
                          </ArkDatePicker.TableRow>
                        ))}
                      </ArkDatePicker.TableBody>
                    </ArkDatePicker.Table>
                  </>
                )}
              </ArkDatePicker.Context>
            </ArkDatePicker.View>
            <ArkDatePicker.View view="year">
              <ArkDatePicker.Context>
                {(datePicker) => (
                  <>
                    <Header />
                    <ArkDatePicker.Table>
                      <ArkDatePicker.TableBody>
                        {datePicker.getYearsGrid({ columns: 4 }).map((years, i) => (
                          <ArkDatePicker.TableRow key={`year_${i}`}>
                            {years.map((year) => (
                              <ArkDatePicker.TableCell key={year.label} value={year.value}>
                                <ArkDatePicker.TableCellTrigger
                                  className={getYearCellClasses(year)}
                                  disabled={isInvalidYear(year.value)}
                                >
                                  {year.label}
                                </ArkDatePicker.TableCellTrigger>
                              </ArkDatePicker.TableCell>
                            ))}
                          </ArkDatePicker.TableRow>
                        ))}
                      </ArkDatePicker.TableBody>
                    </ArkDatePicker.Table>
                  </>
                )}
              </ArkDatePicker.Context>
            </ArkDatePicker.View>
          </ArkDatePicker.Content>
        </ArkDatePicker.Positioner>
      </Portal>
    </ArkDatePicker.Root>
  );
});
