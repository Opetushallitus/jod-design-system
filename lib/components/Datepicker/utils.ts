import { parseDate } from '@ark-ui/react';
import { CalendarDate, DateValue } from '@internationalized/date';
import { clamp } from '../../utils';

export const MIN_YEAR = 1900;
export const MAX_YEAR = 2100;
export const isInvalidYear = (year: number) => year < MIN_YEAR || year > MAX_YEAR;

/** Tries to parse the string value from Datepicker input into CalendarDate */
export const parseInputValue = (value?: string): DateValue | undefined => {
  if (value === '') {
    return;
  }
  if (value) {
    // Matches the format dd.mm.yyyy, should match after selecting a date from the calendar
    const dateRegex = /^([1-9]|0[1-9]|[12]\d|3[01])\.([1-9]|0[1-9]|1[0-2])\.(\d+)$/;
    const match = dateRegex.exec(value);

    if (match) {
      const [date, month, year] = [match[1], match[2], match[3]].map((v) => parseInt(v, 10));
      const fixedYear = year < 100 ? year + 1900 : year;
      const validYear = isInvalidYear(fixedYear) ? clamp(fixedYear, MIN_YEAR, MAX_YEAR) : fixedYear;
      return new CalendarDate(validYear, month, date);
    } else {
      // If no match, try to parse the value using the Ark Datepickers internal parseDate function
      const date = new Date(value);
      if (!isNaN(date.getTime())) {
        const parsed = parseDate(date);
        return new CalendarDate(parsed.year, parsed.month, parsed.day);
      }
    }
  }

  // Fallback to current date
  const date = new Date();
  const day = date.getDate();
  const month = date.getMonth() + 1;
  const year = date.getFullYear();
  return new CalendarDate(year, month, day);
};

/**
 * Makes sure the internal DateValue is within the valid range and changes it if necessary
 * @param date DateValue that the ARK datepicker uses
 * @returns DateValue that is within the valid range
 */
export const verifyCalendarDate = (date: DateValue): CalendarDate => {
  const year = isInvalidYear(date.year) ? clamp(date.year, MIN_YEAR, MAX_YEAR) : date.year;
  const month = clamp(date.month, 1, 12);
  const day = clamp(date.day, 1, new Date(year, month, 0).getDate());
  return new CalendarDate(year, month, day);
};
