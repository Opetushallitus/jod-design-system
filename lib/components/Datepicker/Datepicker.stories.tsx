import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

import { useState } from '@storybook/preview-api';
import { Datepicker } from './Datepicker';

const meta = {
  title: 'Forms/Datepicker',
  component: Datepicker,
  tags: ['autodocs'],
} satisfies Meta<typeof Datepicker>;

export default meta;

type Story = StoryObj<typeof meta>;

const parameters = {
  design: {
    type: 'figma',
    url: 'https://www.figma.com/design/6M2LrpSCcB0thlFDaQAI2J/cx_jod_client?node-id=2217-7791&t=dNDS0896ycg3WZwJ-4',
  },
};

const args = {
  label: 'Valitse päivämäärä',
  placeholder: 'pp.kk.vvvv',
  help: 'Help text',
  onChange: fn(),
};

export const Default: Story = {
  decorators: [
    (Story) => (
      <div className="ds-max-w-[415px]">
        <Story />
      </div>
    ),
  ],
  parameters,
  args,
};

export const WithValue: Story = {
  decorators: [
    (Story) => (
      <div className="ds-max-w-[415px]">
        <Story />
      </div>
    ),
  ],
  parameters,
  render: (args) => {
    const [value, setValue] = useState(args.value);
    return <Datepicker {...args} value={value} onChange={(e) => setValue(e.target.value)} />;
  },
  args: {
    ...args,
    value: '2024-06-01',
  },
};
