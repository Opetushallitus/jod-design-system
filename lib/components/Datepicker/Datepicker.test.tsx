import { act, render, screen } from '@testing-library/react';
import { DateView, DayTableCellState } from '@zag-js/date-picker';
import React from 'react';
import { userEvent } from 'storybook/test';
import { afterEach, describe, expect, it, vi } from 'vitest';
import { Datepicker } from './Datepicker';

describe('Datepicker', () => {
  const label = 'Pick a date dude!';
  const placeholder = 'pp.kk.vvvv';
  const value = '2024-05-01';
  const formattedValue = '1.5.2024';
  const onChange = vi.fn();

  afterEach(() => {
    onChange.mockClear();
  });
  const viewTranslations = {
    day: {
      next: 'Switch to next month',
      view: 'Switch to month view',
      prev: 'Switch to previous month',
    },
    month: {
      next: 'Switch to next year',
      view: 'Switch to year view',
      prev: 'Switch to previous year',
    },
    year: {
      next: 'Switch to next decade',
      view: 'Switch to day view',
      prev: 'Switch to previous decade',
    },
  } as const;

  const translations = {
    nextTrigger: (view: DateView) => viewTranslations[view].next,
    viewTrigger: (view: DateView) => viewTranslations[view].view,
    prevTrigger: (view: DateView) => viewTranslations[view].prev,
    dayCell: (state: DayTableCellState): string => `Choose ${state.formattedDate}`,
    trigger: (open: boolean): string => (open ? 'Close calendar' : 'Open calendar'),
  };

  it('renders correctly', () => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date(2024, 6, 1, 12, 0, 0));
    vi.runAllTimers();

    const { container } = render(
      <Datepicker
        label={label}
        placeholder={placeholder}
        value={value}
        onChange={onChange}
        translations={translations}
      />,
    );
    expect(container.firstChild).toMatchSnapshot();
    vi.useRealTimers();
  });

  it('renders the correct label', () => {
    render(
      <Datepicker
        label={label}
        placeholder={placeholder}
        value={value}
        onChange={onChange}
        translations={translations}
        dataTestId="dp"
      />,
    );
    expect(screen.getByText(label)).toBeInTheDocument();
    expect(screen.getByTestId('dp')).toBeInTheDocument();
    expect(screen.getByTestId('dp-input')).toBeInTheDocument();
    expect(screen.getByTestId('dp-trigger')).toBeInTheDocument();
  });

  it('renders the correct initial value', () => {
    render(
      <Datepicker
        label={label}
        placeholder={placeholder}
        value={value}
        onChange={onChange}
        translations={translations}
      />,
    );
    expect(screen.getByDisplayValue(formattedValue)).toBeInTheDocument();
  });

  it('calls onValueChange when a new date is typed into the input', async () => {
    render(
      <Datepicker
        name="startDate"
        label={label}
        placeholder={placeholder}
        onChange={onChange}
        translations={translations}
      />,
    );
    const input = screen.getByRole('textbox');
    const user = userEvent.setup();
    await act(async () => {
      await user.type(input, '01.06.2024');
      await user.tab();
    });
    expect(onChange).toHaveBeenCalledWith(
      expect.objectContaining({
        target: { name: 'startDate', value: '2024-06-01' },
      }),
    );
  });

  it('can clear the value', async () => {
    const value = {
      date: '2020-01-01',
    };
    const onChangeMock = vi.fn().mockImplementation((event: React.ChangeEvent<HTMLInputElement>) => {
      value.date = event.target.value;
    });
    render(
      <Datepicker
        name="startDate"
        label={label}
        placeholder={placeholder}
        onChange={onChangeMock}
        value={value.date}
        translations={translations}
      />,
    );
    const input = screen.getByRole('textbox');
    const user = userEvent.setup();
    await act(async () => {
      await user.clear(input);
      await user.tab();
    });

    expect(onChangeMock).toHaveBeenCalledWith(
      expect.objectContaining({
        target: { name: 'startDate', value: '' },
      }),
    );

    expect(value.date).toBe('');
  });
});
