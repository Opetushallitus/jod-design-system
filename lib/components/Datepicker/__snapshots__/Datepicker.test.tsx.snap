// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Datepicker > renders correctly 1`] = `
<div
  class="ds:w-full"
  data-part="root"
  data-scope="date-picker"
  data-state="closed"
  dir="ltr"
  id="datepicker::r1:"
>
  <label
    class="ds:mb-4 ds:inline-block ds:align-top ds:text-form-label ds:font-arial ds:text-black"
    data-index="0"
    data-part="label"
    data-scope="date-picker"
    data-state="closed"
    dir="ltr"
    for="datepicker::r1::input:0"
    id="datepicker::r1::label:0"
  >
    Pick a date dude!
  </label>
  <div
    data-part="control"
    data-scope="date-picker"
    dir="ltr"
    id="datepicker::r1::control"
  >
    <div
      class="ds:flex ds:w-full"
    >
      <input
        autocomplete="off"
        autocorrect="off"
        class="ds:w-full ds:rounded-l ds:border-y ds:border-l ds:border-border-gray ds:bg-white ds:p-5 ds:font-arial ds:text-black ds:outline-hidden ds:placeholder:text-inactive-gray ds:placeholder:text-body-md"
        data-index="0"
        data-part="input"
        data-scope="date-picker"
        data-state="closed"
        dir="ltr"
        id="datepicker::r1::input:0"
        placeholder="pp.kk.vvvv"
        spellcheck="false"
        value="1.5.2024"
      />
      <button
        aria-controls="datepicker::r1::content"
        aria-haspopup="grid"
        aria-label="Open calendar"
        class="ds:rounded-r ds:border-y ds:border-r ds:border-border-gray ds:bg-white ds:p-5 ds:text-secondary-gray ds:cursor-pointer"
        data-part="trigger"
        data-scope="date-picker"
        data-state="closed"
        dir="ltr"
        id="datepicker::r1::trigger"
        type="button"
      >
        <svg
          fill="currentColor"
          height="24"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0z"
            fill="none"
          />
          <path
            d="M20 3h-1V1h-2v2H7V1H5v2H4c-1.1 0-2 .9-2 2v16c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zm0 18H4V8h16v13z"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
